### PR TITLE
Tool chaining

### PR DESCRIPTION
**Pull request by cameronfabbri**
- [x] Changed default GPT model to `gpt-4-0125-preview`
- [x] Added functionality to call multiple tools until an answer has been reached.
- [x] Added dispatch messages to main message history
- [x] Updated `parse_response` to only return the first Tool, and to handle newlines better.
- [x] Added more unit tests for `parse_response`

A prompt to test: `Find the method that calculates the running cost, then run the linting tool on the file it's in.`

### CODE REVIEW

File: src/groundcrew/agent.py [hunk 1]
```diff
@@ -113,106 +122,90 @@ def run(self):
 
             self.interact(user_prompt)
 
-    def run_with_prompts(self, prompts: list[str]):
-        """
-        Process a list of user prompts and respond using the chosen tool
-        based on the input.
+    def run_tool(self, parsed_response):
```
Comment (bdzimmer) [inline src/groundcrew/agent.py:125]: type annotation

File: src/groundcrew/agent.py [hunk 2]
```diff
@@ -113,106 +122,90 @@ def run(self):
...
+            print(f'("{parsed_response["Tool query"]}", {tool_args})')
+            print()
 
-        Args:
-            user_prompt (str): The user's input or question.
+        return tool.obj(parsed_response['Tool query'], **tool_args)
 
-        Returns:
-            str: The response from the tool or LLM.
+    def dispatch(self, user_prompt: str) -> str:
```
Comment (bdzimmer) [inline src/groundcrew/agent.py:158]: this no longer returns a str

File: src/groundcrew/agent.py [hunk 3]
```diff
@@ -113,106 +122,90 @@ def run(self):
...
+                color='green'
             )
+            self.spinner.start()
+
+            # Run Tool
+            tool_response = self.run_tool(parsed_select_tool_response)
+            self.spinner.stop()
 
-        return response
+            self.dispatch_messages.append(
```
Comment (bdzimmer) [inline src/groundcrew/agent.py:206]: wrap

File: src/groundcrew/system_prompts.py [hunk 4]
```diff
@@ -87,3 +86,5 @@ def __call__(self, user_prompt: str, parameter_1: str):
         type: str
         required: true
 """
+
+_CHOOSE_TOOL_PROMPT = """Your task is to address a question or command from a user in the ### Question ### seciton. You will do this in a step by step manner by choosing the correct Tool and parameters necessary for this task. The previous Tools used and their results are available to you in the ### Previous Steps ### section. If the ### Previous Steps ### sections is empty, generate a short step by step plan for how you plan to complete the task. Be sure to ask the user for confirmation of your plan, and revise your plan as needed. Only include "Tool:" in your answer if you are choosing a valid Tool available to you. When you have the necessary data to complete your task, respond directly to the user with a summary of the steps taken. Your answer must be in one of the following two formats."""
```
Comment (bdzimmer) [inline src/groundcrew/system_prompts.py:90]: I think this can be deleted

**Review by bdzimmer — APPROVED** [review]
This looks great, just minor style comments. Will be fun to test with eval framework to see if the questions we've collected so far still work.

File: src/groundcrew/utils.py [hunk 5]
```diff
@@ -81,10 +81,8 @@ def build_llm_chat_client(
         client = openaiapi.get_openaiai_client()
         chat_session = openaiapi.start_chat(model, client)
 
-        def chat(messages: list[Message]) -> str:
-            response = chat_session(messages)
-            messages.append(response)
-            return response.content
+        def chat(messages: list[Message]) -> Message:
+            return chat_session(messages)
```
Comment (jmpounders) [inline src/groundcrew/utils.py:85]: Thanks for fixing my mistake 👍

**Review by jmpounders — APPROVED** [review]

**Comment by jmpounders** [comment]
Nice work.  The refactor looks great.

**Pull request merged**